From 469249ba9e1fd6f435d73a3c43cadbaca0535c81 Mon Sep 17 00:00:00 2001
From: Niko Tyni <ntyni@debian.org>
Date: Sat, 13 Sep 2014 21:35:43 +0300
Subject: [PATCH] Test cases for needs_recommends bug

If there are two tests with the same dependencies but different
needs_recommends restrictions, the test bed should not be re-used
between them.
---
 tests/adt-run | 34 ++++++++++++++++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/tests/adt-run b/tests/adt-run
index 2187e3e..61b99fe 100755
--- a/tests/adt-run
+++ b/tests/adt-run
@@ -1952,6 +1952,40 @@ class SchrootRunner(AdtTestCase):
         self.assertIn('Unpacking cpp-doc', out)
         self.assertIn('built script OK', out)
 
+    def test_needs_recommends2(self):
+        '''needs-recommends restriction in later of two tests'''
+
+        # check that testpkg's Recommends: gets installed
+        # even if there's a preceding test with the same dependencies
+        # but without needs-recommends
+        p = self.build_src('Tests: pass1\n\nTests: pass2\nRestrictions: needs-recommends',
+                           {'pass1': '#!/bin/sh -e\n'
+                            'test ! -e /usr/share/doc/cpp-doc/copyright ||'
+                            'echo "cpp-doc installed too early!" >&2',
+                            'pass2': '#!/bin/sh -e\n'
+                            'test -e /usr/share/doc/cpp-doc/copyright ||'
+                            'echo "cpp-doc not installed!" >&2'})
+
+        (code, out, err) = self.adt_run(['--unbuilt-tree=' + p], [self.schroot_name])
+        self.assertEqual(code, 0, out + err)
+
+    def test_needs_recommends3(self):
+        '''needs-recommends restriction in former of two tests'''
+
+        # check that testpkg's Recommends: isn't installed without
+        # needs-recommends even if there's a preceding test with the
+        # same dependencies but with needs-recommends
+        p = self.build_src('Tests: pass1\nRestrictions: needs-recommends\n\nTests: pass2\n',
+                           {'pass1': '#!/bin/sh -e\n'
+                            'test -e /usr/share/doc/cpp-doc/copyright ||'
+                            'echo "cpp-doc not installed!" >&2',
+                            'pass2': '#!/bin/sh -e\n'
+                            'test ! -e /usr/share/doc/cpp-doc/copyright ||'
+                            'echo "cpp-doc still installed!" >&2'})
+
+        (code, out, err) = self.adt_run(['--unbuilt-tree=' + p], [self.schroot_name])
+        self.assertEqual(code, 0, out + err)
+
 
 @unittest.skipUnless('ADT_TEST_LXC' in os.environ,
                      'Set $ADT_TEST_LXC to an existing container')
-- 
2.1.0

